### ISSUE TITLE
Signed integer overflow not detected with -fsanitize=undefined -O1 optimization level

### ISSUE THREAD

**Issue by wangbo15** [llvm/llvm-project#121053]
The following program shows an issue where signed integer overflow is not detected when compiled with `Clang` using the option `-fsanitize=undefined -O1`. However, without `-O1`, the runtime error is correctly reported as an overflow.
Additionally, `GCC` rejects this by both `-fsanitize=undefined` and `-fsanitize=undefined -O1`.
Please see https://godbolt.org/z/j5K3PPac8

```
#include <stdio.h>
unsigned a;
class b {
public:
  char d[25][25];
} *e = new b;

void func(b *) {
    for (int y = 3;
           y < -5870024400429537026 * (e->d[7][0] / 012) - 6706695272850477542;
           y += 4) {
        a = !0;
      }
}

int main() {
    for (int i; i < 25; ++i)
        for (int j = 0; j < 5; ++j)
            e->d[i][j] = 033;

    func(e);
    printf("%u\n", a);
}
```

**Comment by dtcxzyw** [comment]
Bisected to LICM.

**Comment by dtcxzyw** [comment]
The original code contains UB. MemorySanitizer reports `use-of-uninitialized-value`:
```
> clang++ -fsanitize=memory -ggdb test.cpp && ./a.out
==2812473==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x5555556232f5 in main /home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/test.cpp:17:5
    #1 0x7ffff7829d8f in __libc_start_call_main csu/../sysdeps/nptl/libc_start_call_main.h:58:16
    #2 0x7ffff7829e3f in __libc_start_main csu/../csu/libc-start.c:392:3
    #3 0x555555587354 in _start (/home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/a.out+0x33354) (BuildId: c2bbf0cf9731696d98c4932e0371b6c539803ff0)

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/dtcxzyw/WorkSpace/Projects/compilers/LLVM/llvm-build/test.cpp:17:5 in main
Exiting
```
After fixing this UB, ubsan works as expected: https://godbolt.org/z/5zdE8P371
```
#include <stdio.h>
unsigned a;
class b {
public:
  char d[25][25];
} *e = new b;

void func(b *) {
    for (int y = 3;
           y < -5870024400429537026 * (e->d[7][0] / 012) - 6706695272850477542;
           y += 4) {
        a = !0;
      }
}

int main() {
    for (int i = 0; i < 25; ++i) // Initialize i with 0
        for (int j = 0; j < 5; ++j)
            e->d[i][j] = 033;

    func(e);
    printf("%u\n", a);
}
```
